### PR TITLE
Issue #17136: Changes to choice method of Random module to support sampling on N-Dimensions

### PR DESCRIPTION
**Pull request by prashanth018**
Changes to choice method of Random module to support sampling on N-Dimensional Array/Domain. Also have added few test cases (need to refine). Existing tests run without any issues.

### CODE REVIEW

**Comment by bradcray** [comment]
Hi @prashanth018 : I'm cleaning out old pull requests, and am noticing that this one has received no attention.  Did you reach out to the team to ask for a review?  (we often don't notice PRs unless someone brings them to our attention due to the number that are open at any given time).  Thanks for any information.

**Comment by prashanth018** [comment]
Hi @bradcray, Yes I did reach out and it was apparent and understandable that most of the reviewers were busy with the release. It would be good if one of you could take a look at it. Also, given that I'm already working on my GSoC project, there might be few delays in merging this issue. Thanks. :)

File: modules/standard/Random.chpl [hunk 1]
```diff
@@ -284,18 +279,45 @@ module Random {
...
+  proc getIndicesDefinition(shape: domain, param rank) {
+    if rank == 1 {
+      var indices: [shape] int;
+      return indices;
+    }
+    else {
+      var indices: [shape] rank * int;
+      return indices;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (e-kayrakli) [reply modules/standard/Random.chpl:310]: Reading the code a bit more, I'd remove both of these helpers, and add a type function helper like

```chapel
private proc _choiceIndexTypeHelper(X: domain) type {
  return if X.rank == 1 then int else X.rank*int;
}
```

(I'm not a big fan of the preceding `_`, but that's how similar functions are named, and let's be consistent)

And then change the array declarations where you're currently using these functions as

```chapel
var a: [b] _choiceIndexTypeHelper(X);
```

and such.

File: modules/standard/Random.chpl [hunk 2]
```diff
@@ -352,17 +375,15 @@ module Random {
...
     }
-    
+
     if prob.size == 0 then
       throw new owned IllegalArgumentError('choice() prob array cannot be empty');
 
-    const low = X.alignedLow,
-          stride = abs(X.stride);
-    ref P = prob.reindex(0..<X.size);
+    ref P = reshape(prob, {0..<X.size});
```
Comment (e-kayrakli) [inline modules/standard/Random.chpl:386]: I don't have a problem with this, but, is there any reason you are not keeping the previous idiom? (`prob.reindex(bla)`)

File: modules/standard/Random.chpl [hunk 3]
```diff
@@ -1134,17 +1155,17 @@ module Random {
...
+     :arg x: a N-D dom with values that will be sampled from.
      :arg size: An optional integral value specifying the number of elements to
                 choose, or a domain specifying the dimensions of the
                 sampled array to be filled, otherwise a single element will be
                 chosen.
      :arg replace: an optional ``bool`` specifying whether or not to sample with
                    replacement, i.e. elements will only be chosen up to one
                    time when ``replace=false``.
-     :arg prob: an optional 1-D array that contains probabilities of choosing
+     :arg prob: an optional N-D array that contains probabilities of choosing
```
Comment (e-kayrakli) [inline modules/standard/Random.chpl:1168]: I think we can remove dimensionality from all these statements. `an optional array` is sufficient, to me.

File: test/library/standard/Random/RandomStreamInterface/choiceTestUtils.chpl [hunk 4]
```diff
@@ -48,20 +52,20 @@ proc test(stream, X, size:?sizeType=none, replace=true, prob:?probType=none, tri
...
       actualRatios[k] = v / (trials*m): real;
   }
-  
+
   var dom: domain(1) = {1..X.size};
   var ones: [dom] real = 1;
 
   var probabilities = if isNothingType(prob.type) then ones
-                      else prob.reindex(1..X.size);
+                      else reshape(prob, {1..X.size});
```
Comment (e-kayrakli) [inline test/library/standard/Random/RandomStreamInterface/choiceTestUtils.chpl:68]: I am seeing a similar change here, so there must be something that I am missing that makes this necessary.

File: test/library/standard/Random/RandomStreamInterface/choiceTestUtils.chpl [hunk 5]
```diff
@@ -6,7 +6,11 @@ proc test(stream, X, size:?sizeType=none, replace=true, prob:?probType=none, tri
   var counts;
   if isArrayValue(X) {
     counts = new map(X.eltType, int);
-  } else {
+  }
+  else if X.rank > 1 {
```
Comment (e-kayrakli) [inline test/library/standard/Random/RandomStreamInterface/choiceTestUtils.chpl:10]: Could you at least add a comment here saying that if we're hitting this `if`, `X` must be domain? You can also make the code more self-documenting by 

```chapel
else if isDomainValue(x) {
  if X.rank > 1 {

  }
  else {

  }
}
else {
  compilerError("Unexpected type");
}

File: modules/standard/Random.chpl [hunk 6]
```diff
@@ -62,6 +62,7 @@
 module Random {
 
   public use RandomSupport;
+  use ChapelArray;
```
Comment (e-kayrakli) [inline modules/standard/Random.chpl:65]: Why did we need this?

**Review by e-kayrakli — CHANGES_REQUESTED** [review]
Thanks @prashanth018! Please see my in-line comments.

**Comment by e-kayrakli** [comment]
@prashanth018 -- as you've been silent here and in your other PRs, I am going to go ahead and close this. You're more than welcome to reopen if/when you want to revisit this.

**Pull request closed without merging**